core/drivers/crypto/stm32/stm32_hash.c

```diff
@@ -0,0 +1,909 @@
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	dst_buf = dst->remain.buf;
+	dst_csr = dst->csr;
+	memcpy(dst, src, sizeof(*dst));
+	dst->remain.buf = dst_buf;
+	dst->csr = dst_csr;
+
+	memcpy(dst->remain.buf, src->remain.buf, dst->remain.len);
+	get_save_registers(dst, &nb_reg, &first, &hmac_nb_reg, &hmac_first);
+	memcpy(dst->csr, src->csr, (nb_reg + hmac_nb_reg) * sizeof(uint32_t));
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_hash_alloc(struct stm32_hash_context *c,
+			    enum stm32_hash_mode mode,
+			    enum stm32_hash_algo algo)
+{
+	size_t nb_reg = 0;
+	size_t first = 0;
+	size_t hmac_nb_reg = 0;
+	size_t hmac_first = 0;
+
+	assert(c);
+
+	/* Check if initialized */
+	if (!stm32_hash)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	c->dev = stm32_hash;
+	c->mode = mode;
+	c->algo = algo;
+
+	switch (algo) {
+	case STM32_HASH_MD5:
+		if (!(c->dev->pdata.compat->caps & CAPS_MD5))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = MD5_DIGEST_U32;
+		c->block_size = MD5_BLOCK_SIZE;
+		c->save_mode = SAVE_SMALL;
+		break;
+	case STM32_HASH_SHA1:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA1))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA1_DIGEST_U32;
+		c->block_size = SHA1_BLOCK_SIZE;
+		c->save_mode = SAVE_SMALL;
+		break;
+	case STM32_HASH_SHA224:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA2_224))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA224_DIGEST_U32;
+		c->block_size = SHA224_BLOCK_SIZE;
+		c->save_mode = SAVE_SMALL;
+		break;
+	case STM32_HASH_SHA384:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA2_384))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA384_DIGEST_U32;
+		c->block_size = SHA384_BLOCK_SIZE;
+		c->save_mode = SAVE_BIG;
+		break;
+	case STM32_HASH_SHA512:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA2_512))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA512_DIGEST_U32;
+		c->block_size = SHA512_BLOCK_SIZE;
+		c->save_mode = SAVE_BIG;
+		break;
+	case STM32_HASH_SHA3_224:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_224_DIGEST_U32;
+		c->block_size = SHA3_224_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_256:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_256_DIGEST_U32;
+		c->block_size = SHA3_256_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_384:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_384_DIGEST_U32;
+		c->block_size = SHA3_384_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_512:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_512_DIGEST_U32;
+		c->block_size = SHA3_512_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	/* Default selected algo is SHA256 */
```

Comment:
This seems strange here. I think you can remove this inline comment and move the case between SHA224 and SAH384 cases.